### PR TITLE
Add Style/Round cop

### PR DESCRIPTION
**Pull request by cyberdelia**
Add Style/Round cop, promoting using ``Float#floor`` or ``Float#ceil`` rather than using ``Float#round``.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

### CODE REVIEW

**Comment by backus** [comment]
If this is related to `#round` behavior changing in 2.4 then FYI they went back on the decision: https://bugs.ruby-lang.org/issues/12958

> OK, the default behavior should be kept unchanged.
>
> Matz.

**Comment by cyberdelia** [comment]
@backus This is unrelated.

File: spec/rubocop/cop/style/round_spec.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      inspect_source(cop, '6.75.round')
+      expect(cop.offenses.size).to eq(1)
+    end
+  end
+
+  context 'with ceil' do
+    after(:each) { expect(cop.offenses).to be_empty }
+
+    it 'does not register an offense for an invocation with args' do
+      inspect_source(cop, '6.75.ceil')
```
Comment (NobodysNightmare) [inline spec/rubocop/cop/style/round_spec.rb]: This is not an invocation with args. It is the same source like the "without" spec.

**Comment by pocke** [comment]
In my opinion, projects that need this cop are limited.
I recommend Querly.  https://github.com/soutaro/querly
This tool provides checking ruby code with project local rule. I think Querly is more suitable for this use than RuboCop. What do you think?

**Comment by Drenmi** [comment]
> This tool provides checking ruby code with project local rule. I think Querly is more suitable for this use than RuboCop. What do you think?

That is a great share @pocke! 😀 

My biggest contention with this is that it changes semantics. This could have some pretty disastrous results if the author actually intended to use `#round`. (I can't see why they would, if they didn't. 🤔)

**Comment by cyberdelia** [comment]
Maybe it's a bit extreme as is, I make changes to only catch calls to ``round`` or ``round(0)```, which effectively are better replaced by either ``floor`` or ``ceil``.

File: lib/rubocop/cop/style/round.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for the use of the round method.
```
Comment (bbatsov) [inline lib/rubocop/cop/style/round.rb]: The documentation here should be significantly extended and include code examples. See other cops for reference.

**Comment by bbatsov** [comment]
I'm guessing that with the narrower scope this cop makes sense. (btw, this would also make a good addition to the Ruby Style Guide). One more thing - the cop needs auto-correct support.

**Comment by cyberdelia** [comment]
@bbatsov I'm not sure how to automate the choice between ``floor`` and ``ceil`` for auto-correct support 😕

**Comment by bbatsov** [comment]
Of course. Silly me. :-) On second thought - why is using `round` worse than `floor` or `ceil`? `round` does exactly what it's supposed to - round down or up depending on the exact value. Seems to me there will always be situations where `round` is exactly what you need, so probably this check is not a good idea at all. Obviously using `floor` or `ceil` is better if you really need them, but I doubt you always do.

**Comment by Drenmi** [comment]
> Seems to me there will always be situations where round is exactly what you need, so probably this check is not a good idea at all.

For a second I was wondering if I was missing something here ... 😅 

This cop could only reliably tell you to use `#floor` or `#ceil` if you call `#round` on a literal, but if you do that, you can just round it in your head before typing it out. Maybe, possibly, you could find constant receivers in the same scope and make some recommendation, but it seems a bit far fetched. 🤔

**Comment by bbatsov** [comment]
Guess we can close this PR.

@cyberdelia Thanks for your efforts. Don't be discouraged by the fact we've opted to reject the proposed feature.

**Pull request closed without merging**